src/dune_rules/ctypes/ctypes_rules.ml

```diff
@@ -216,33 +216,54 @@ let build_c_program ~foreign_archives_deps ~sctx ~dir ~source_files ~scope
     in
     deps
   in
-  let build =
-    let absolute_path_hack p =
-      (* These normal path builder things construct relative paths like
-         _build/default/your/project/file.c but before dune runs gcc it actually
-         cds into _build/default, which fails, so we turn them into absolutes to
-         hack around it. *)
-      Path.relative (Path.build dir) p |> Path.to_absolute_filename
-    in
-    let action =
-      let open Action_builder.O in
-      let* include_args = Resolve.Memo.read include_args in
-      let* base_args = with_user_and_std_flags in
-      deps
-      >>> Action_builder.map cflags ~f:(fun cflags_args ->
-              let source_files = List.map source_files ~f:absolute_path_hack in
-              let output = absolute_path_hack output in
-              let args =
-                base_args @ cflags_args @ include_args @ source_files
-                @ [ "-o"; output ]
-              in
-              Action.run exe args)
-    in
-    Action_builder.with_file_targets action
-      ~file_targets:[ Path.Build.relative dir output ]
+  let action =
+    if version >= (0, 3) then
+      let args : _ Command.Args.t list =
+        [ Expand (fun ~dir:_ -> with_user_and_std_flags)
+        ; Expand (fun ~dir:_ -> cflags)
+        ; Expand (fun ~dir:_ -> Resolve.Memo.read include_args)
```

Comment:
Could you explain why we need the `Expand` here? It seems strange to use it without using the directory argument.

Reply:
ah, good point. I was looking for a way to use an `Action_builder.t` in `Command.Args.t` but I had missed that `Dyn` would work better. I fixed it by using `Dyn` instead.